### ISSUE TITLE
Thunderbird fails with `Error: queue file write error` when sending mail

### ISSUE THREAD

**Issue by Encephala** [docker-mailserver/docker-mailserver#3519]
### 📝 Preliminary Checks

- [X] I tried searching for an existing issue and followed the [debugging docs](https://docker-mailserver.github.io/docker-mailserver/latest/config/debugging/) advice, but still need assistance.

### 👀 What Happened?

I had DMS set up and working with local users, succesfully sent an email to my personal email address. I used a MUA (Mozilla Thunderbird) and am able to authenticate with local users and send email, all works perfectly!

That was just for testing, in the end we need to use LDAP for authentication. I set all the LDAP settings, set `ACCOUNT_PROVISIONER=LDAP`, `ENABLE_LDAP=1` and enabled SASLauthd using `rimap`. When I execute into the container and run `doveadm auth test`, authentication to the LDAP server is succesful.

Back to mozilla and IMAP authentication is succesful. However, when I try to send an email, Thunderbird says that the mail server reported an error, namely 'Error: queue file write error'. Strangely, I don't see any such error in the logs, only warnings. See the logs below.

Weirdly, the logs only report warnings, no errors. Googling any of the warnings does not give relevant results. 
My problem is that it is unclear which of these (if any) are the cause of the email failing to send, and the warning Thunderbird reported doesn't give much insight either. 
I'd gladly believe that this is user misconfiguration, but since there is no clear error logging, it feels like something is going wrong inside DMS, hence this bug report.
It's unclear to me from the documentation if SASLauthd is even necessary, I can't really find what postfix uses for authentication. Perhaps that's where things are going wrong.

As a footnote, since we are currently migrating our mailserver from a manual installation to DMS, so we have two parallel MX-records on the same domain that point to two different servers. This shouldn't be an issue for sending emails, but I'm no email expert so I'm not sure, so I figured I'd mention it anyways.

### 👟 Reproduction Steps

Use LDAP authentication, with SASLauthd configured to authenticate Dovecot (`rimap`). 

### 🐋 DMS Version

v12.1.0

### 💻 Operating System and Architecture

Ubuntu 22.04 on a cloud-provided VPS

### ⚙️ Container configuration files

```yml
services:
  mailserver:
    image: docker.io/mailserver/docker-mailserver:12.1.0

    hostname: mail2.sintjansbrug.nl

    env_file: mailserver.env

    ports:
      - "25:25"    # SMTP  (explicit TLS => STARTTLS)
      - "143:143"  # IMAP4 (explicit TLS => STARTTLS)
      - "465:465"  # ESMTP (implicit TLS)
      - "587:587"  # ESMTP (explicit TLS => STARTTLS)
      - "993:993"  # IMAP4 (implicit TLS)

    volumes:
      - "/mnt/bigstorage/docker-mailserver/data/:/var/mail/"
      - "/mnt/bigstorage/docker-mailserver/mail-state/:/var/mail-state/"
      - "/mnt/bigstorage/docker-mailserver/logs/:/var/log/mail/"
      - "./config/:/tmp/docker-mailserver/"
      # https://github.com/docker-mailserver/docker-mailserver/issues/475#issuecomment-899652084
      - "./dovecot-local.conf:/etc/dovecot/conf.d/99-local.conf"
      - "/etc/localtime:/etc/localtime:ro"

    restart: unless-stopped
    stop_grace_period: 1m
```

### 📜 Relevant log output

```Text
Aug 30 16:24:39 mail2 postfix/smtps/smtpd[978]: warning: dict_ldap_open: /etc/postfix/ldap-aliases.cf: Fixed query_filter  is probably useless
Aug 30 16:24:39 mail2 postfix/smtps/smtpd[978]: warning: dict_ldap_open: /etc/postfix/ldap-groups.cf: Fixed query_filter  is probably useless
Aug 30 16:24:39 mail2 postfix/smtps/smtpd[978]: connect from <hostname>[<ip-address>]
Aug 30 16:24:39 mail2 postfix/smtps/smtpd[978]: Anonymous TLS connection established from <hostname>[<ip-address>]: TLSv1.3 with cipher TLS_AES_256_GCM_SHA384 (256/256 bits) key-exchange X25519 server-signature RSA-PSS (2048 bits) server-digest SHA256
Aug 30 16:24:49 mail2 dovecot: imap-login: Login: user=<username@domain.com>, method=PLAIN, rip=127.0.0.1, lip=127.0.0.1, mpid=983, secured, session=<JLmpsCQEetJ/AAAB>
Aug 30 16:24:49 mail2 dovecot: imap(username@domain.com)<983><JLmpsCQEetJ/AAAB>: Disconnected: Logged out in=16 out=496 deleted=0 expunged=0 trashed=0 hdr_count=0 hdr_bytes=0 body_count=0 body_bytes=0
Aug 30 16:24:49 mail2 postfix/trivial-rewrite[984]: warning: dict_ldap_open: /etc/postfix/ldap-aliases.cf: Fixed query_filter  is probably useless
Aug 30 16:24:49 mail2 postfix/trivial-rewrite[984]: warning: dict_ldap_open: /etc/postfix/ldap-groups.cf: Fixed query_filter  is probably useless
Aug 30 16:24:49 mail2 postfix/trivial-rewrite[984]: warning: dict_ldap_open: /etc/postfix/ldap-domains.cf: Fixed query_filter  is probably useless
Aug 30 16:24:49 mail2 postfix/trivial-rewrite[984]: warning: dict_ldap_lookup: Search error 2: Protocol error
Aug 30 16:24:49 mail2 postfix/trivial-rewrite[984]: warning: virtual_alias_domains: ldap:/etc/postfix/ldap-aliases.cf: table lookup problem
Aug 30 16:24:49 mail2 postfix/trivial-rewrite[984]: warning: virtual_alias_domains lookup failure
Aug 30 16:24:49 mail2 postfix/trivial-rewrite[984]: warning: dict_ldap_lookup: Search error 2: Protocol error
Aug 30 16:24:49 mail2 postfix/trivial-rewrite[984]: warning: virtual_alias_domains: ldap:/etc/postfix/ldap-aliases.cf: table lookup problem
Aug 30 16:24:49 mail2 postfix/trivial-rewrite[984]: warning: virtual_alias_domains lookup failure
Aug 30 16:24:49 mail2 postfix/sender-cleanup/cleanup[985]: warning: dict_ldap_open: /etc/postfix/ldap-aliases.cf: Fixed query_filter  is probably useless
Aug 30 16:24:49 mail2 postfix/sender-cleanup/cleanup[985]: warning: dict_ldap_open: /etc/postfix/ldap-groups.cf: Fixed query_filter  is probably useless
Aug 30 16:24:49 mail2 postfix/smtps/smtpd[978]: D391AE0004E: client=<hostname>[<ip-address>], sasl_method=PLAIN, sasl_username=username@domain.com
Aug 30 16:24:49 mail2 postfix/sender-cleanup/cleanup[985]: warning: dict_ldap_lookup: Search error 2: Protocol error
Aug 30 16:24:49 mail2 postfix/sender-cleanup/cleanup[985]: warning: ldap:/etc/postfix/ldap-aliases.cf lookup error for "recipient@hotmail.com"
Aug 30 16:24:49 mail2 postfix/sender-cleanup/cleanup[985]: warning: D391AE0004E: virtual_alias_maps map lookup problem for recipient@hotmail.com -- message not accepted, try again later
```
```


### Improvements to this form?

_No response_

**Comment by polarathene** [comment]
**TL;DR:**
- At a glance it looks like there is something misconfigured with your LDAP config files for Postfix. You should share the LDAP ENV in your `mailserver.env` file. The logging is rather clear on this as a failure point.
- `ENABLE_SASLAUTHD=1` is not required for LDAP. It provides alternative SASL provider that supports `ldap` and `rimap` mechanisms (auth backends).
- Mail submission is via ports 465 or 587. Thunderbird is connecting to Postfix for this, Dovecot has nothing to do it with it, beyond as the default auth provider (_or indirectly through SASLAuthd using `rimap`_). Postfix delegates auth to a SASL provider.
- LDAP while supported in DMS for years has barely had proper maintenance support. It's being improved for the next release and the docs should also see revision.

---

> However, when I try to send an email, Thunderbird says that the mail server reported an error, namely 'Error: queue file write error'

Mail submission is through Postfix. So Thunderbird is connecting to port 587 or 465 to perform this.

> That was just for testing, in the end we need to use LDAP for authentication. I set all the LDAP settings, set `ACCOUNT_PROVISIONER=LDAP`, `ENABLE_LDAP=1` and enabled SASLauthd using `rimap`. When I execute into the container and run `doveadm auth test`, authentication to the LDAP server is succesful.

`ENABLE_LDAP=1` is not necessary, docs should be clear about this that `ACCOUNT_PROVISIONER=LDAP` is the new equivalent. This will enable LDAP config in both Postfix and Dovecot. However what has not been documented well (_and current maintainers unaware for quite some time due to lack of LDAP experience_), was Dovecot is configured for authentication through LDAP (_IMAP/POP3, and anything else related to users with mail storage (`/var/mail`) that Dovecot manages_). Postfix auth is still delegated to SASL (by default Dovecot).

> It's unclear to me from the documentation if SASLauthd is even necessary, I can't really find what postfix uses for authentication. Perhaps that's where things are going wrong.

Postfix in DMS is configured to use SASL for auth with Dovecot as the default SASL provider.

SASLAuthd was contributed at a later date by a user that replaced Dovecot with a different service, requiring Postfix to authenticate through a different SASL provider service. Initially this for the LDAP mechanism, but support later expanded to `rimap` for auth via an IMAP login service instead (_such as what Dovecot provides_).

> `env_file: mailserver.env`

This isn't too helpful. ENV is important to understand your DMS configuration better. It's better to share the ENV you've explicitly changed (_minus any sensitive data like secrets_).

---

I don't know about the error message you've got, but `docker logs container_name` / `docker compose logs mailserver` should include the Postfix logs which should show some failure happening there, possibly with the authentication step.

> `warning: dict_ldap_open: /etc/postfix/ldap-domains.cf: Fixed query_filter  is probably useless`

What is your query filter?

> ```
> warning: dict_ldap_lookup: Search error 2: Protocol error
> warning: virtual_alias_domains: ldap:/etc/postfix/ldap-aliases.cf: table lookup problem
> warning: virtual_alias_domains lookup failure
> ```

It doesn't look like your LDAP config is setup correctly. Even if Dovecot can authenticate successfully, Postfix is clearly having trouble with it's config to do various LDAP queries to perform its job. Probably the main cause of the failure you're experiencing.

> ```
> D391AE0004E: client=<hostname>[<ip-address>], sasl_method=PLAIN, sasl_username=username@domain.com
> warning: dict_ldap_lookup: Search error 2: Protocol error
> warning: ldap:/etc/postfix/ldap-aliases.cf lookup error for "recipient@hotmail.com"
> warning: D391AE0004E: virtual_alias_maps map lookup problem for recipient@hotmail.com -- message not accepted, try again later
> ```

Here the log starts with `smtpd` (you've logged in and you're sending a mail to `recipient@hotmail.com`, but the LDAP lookups are failing when it attempts to check if this address has a match in `virtual_alias_maps` (_managed by `ldap-aliases.cf`_). The remaining logs were from the Postfix `cleanup` service and we can see that shared session ID `D391AE0004E` with the `smtpd` SASL login.

---

Additionally, right now Postfix is configured to append `$mydomain` if the login name is missing `@`, so if you auth with Thunderbird for `username` with your `hostname: mail2.sintjansbrug.nl`, it'd be treated as `username@sintjansbrug.nl`. 

However, SASLAuthd `rimap` uses `-r` when starting the saslauthd process, while the alternative `ldap` (_which you're not using_) presently does not. When `-r` is not used, the "realm" or domain-part is stripped from the LDAP query during auth, even when using `myAttribute=%u`, effectively making it `%U`. That can lead to some difficult to troubleshoot gotchas, which perhaps you ran into if you tried `ldap` before choosing `rimap`? I am planning to address this for consistency by also having `ldap` start the `saslauthd` process with the `-r` option.

That implicit Postfix behaviour with `$mydomain` is valid elsewhere, it'll be used to create your "username" as understood by Postfix. So for features like `SPOOF_PROTECTION=1` it'll be doing queries to check for results with that "username", while the query input for other LDAP table configs Postfix is configured with will often be the sender or recipient address (_or for the `ldap-domains.cf`, the domain-part_).

---

The LDAP support in DMS has lacked a maintainer for years, but [I've recently been going through it to try tidy up our support](https://github.com/docker-mailserver/docker-mailserver/issues/3517) for the next major release (_expected to be released before 2024_), but I'm also likely to introduce some breaking changes.

If it's helpful, you can [find our LDAP tests here](https://github.com/docker-mailserver/docker-mailserver/blob/3d5f6aeec40ac1c719becdb031fe306f33bcfcc7/test/tests/serial/mail_with_ldap.bats). We test against OpenLDAP with [a simple config via LDIF files](https://github.com/docker-mailserver/docker-mailserver/tree/3d5f6aeec40ac1c719becdb031fe306f33bcfcc7/test/config/ldap/openldap), and run like so:

https://github.com/docker-mailserver/docker-mailserver/blob/3d5f6aeec40ac1c719becdb031fe306f33bcfcc7/test/tests/serial/mail_with_ldap.bats#L23-L33

Some of the test-cases cover sending mail (locally to DMS), and they use auth to Postfix through port 465.

**Comment by Encephala** [comment]
Thanks for your extensive answer! I've found our issue:

> What is your query filter?

It's an empty filter. Our LDAP only has users and groups, but no aliases or domains specified, because we only have one domain and will use GNU Mailman to handle aliases and mailing lists. We also hence have no use for a group filter. 

These filters were therefore not set, i.e. they were an empty string.
I've changed these filters from empty string to `(|)` based on [the sample LDAP configuration](https://docker-mailserver.github.io/docker-mailserver/latest/config/advanced/auth-ldap/), and SMTP authentication now works. 
I did also set `ENABLE_SASLAUTHD=0` again, but I don't think that made a difference; if I'm understanding this right, SASLAUTHD wasn't being used in the first place?
I tested with these filters set back to an empty string and got the same authentication issues again, so I can confirm that this was indeed our problem. 

Perhaps it's a good idea to set `(|)` as the default values for each of these four query filters in the sample env file?

For future reference, the logs still report all of the same things as before:

```
Sep  1 12:30:47 mail2 postfix/smtps/smtpd[973]: warning: dict_ldap_open: /etc/postfix/ldap-aliases.cf: Fixed query_filter (|) is probably useless
Sep  1 12:30:47 mail2 postfix/smtps/smtpd[973]: warning: dict_ldap_open: /etc/postfix/ldap-groups.cf: Fixed query_filter (|) is probably useless
Sep  1 12:30:51 mail2 postfix/smtps/smtpd[975]: warning: dict_ldap_open: /etc/postfix/ldap-aliases.cf: Fixed query_filter (|) is probably useless
Sep  1 12:30:51 mail2 postfix/smtps/smtpd[975]: warning: dict_ldap_open: /etc/postfix/ldap-groups.cf: Fixed query_filter (|) is probably useless
Sep  1 12:30:51 mail2 postfix/smtps/smtpd[975]: connect from <domain>[<ip>]
Sep  1 12:30:52 mail2 postfix/smtps/smtpd[975]: Anonymous TLS connection established from <domain>[<ip>]: TLSv1.3 with cipher TLS_AES_256_GCM_SHA384 (256/256 bits) key-exchange ECDHE (P-256) server-signature RSA-PSS (2048 bits) server-digest SHA256
Sep  1 12:30:52 mail2 postfix/sender-cleanup/cleanup[978]: warning: dict_ldap_open: /etc/postfix/ldap-aliases.cf: Fixed query_filter (|) is probably useless
Sep  1 12:30:52 mail2 postfix/sender-cleanup/cleanup[978]: warning: dict_ldap_open: /etc/postfix/ldap-groups.cf: Fixed query_filter (|) is probably useless
Sep  1 12:30:52 mail2 postfix/smtps/smtpd[975]: 9ECF9E0004C: client=<domain>[<ip>], sasl_method=PLAIN, sasl_username=email@domain.com
Sep  1 12:30:52 mail2 postfix/sender-cleanup/cleanup[978]: 9ECF9E0004C: message-id=<AJYA0S.BEBA7LXKKXBE2@domain.com>
Sep  1 12:30:53 mail2 opendkim[829]: 9ECF9E0004C: DKIM-Signature field added (s=mail, d=domain.com)
Sep  1 12:30:53 mail2 postfix/qmgr[944]: 9ECF9E0004C: from=<email@domain.com>, size=240, nrcpt=1 (queue active)
Sep  1 12:30:53 mail2 postfix/smtpd-amavis/smtpd[981]: warning: dict_ldap_open: /etc/postfix/ldap-aliases.cf: Fixed query_filter (|) is probably useless
Sep  1 12:30:53 mail2 postfix/smtpd-amavis/smtpd[981]: warning: dict_ldap_open: /etc/postfix/ldap-groups.cf: Fixed query_filter (|) is probably useless
Sep  1 12:30:53 mail2 postfix/smtpd-amavis/smtpd[981]: connect from localhost[127.0.0.1]
Sep  1 12:30:53 mail2 postfix/smtps/smtpd[975]: disconnect from <domain>[<ip>] ehlo=1 auth=1 mail=1 rcpt=1 data=1 quit=1 commands=6
Sep  1 12:30:53 mail2 postfix/cleanup[982]: warning: dict_ldap_open: /etc/postfix/ldap-aliases.cf: Fixed query_filter (|) is probably useless
Sep  1 12:30:53 mail2 postfix/cleanup[982]: warning: dict_ldap_open: /etc/postfix/ldap-groups.cf: Fixed query_filter (|) is probably useless
Sep  1 12:30:53 mail2 postfix/smtpd-amavis/smtpd[981]: 4956DE0004E: client=localhost[127.0.0.1]
Sep  1 12:30:53 mail2 postfix/cleanup[982]: 4956DE0004E: message-id=<AJYA0S.BEBA7LXKKXBE2@domain.com>
Sep  1 12:30:53 mail2 postfix/qmgr[944]: 4956DE0004E: from=<email@domain.com>, size=1321, nrcpt=1 (queue active)
Sep  1 12:30:53 mail2 amavis[952]: (00952-01) Passed CLEAN {RelayedOpenRelay}, [<ip>]:59230 [<ip>] <email@domain.com> -> <recipient@hotmail.com>, Queue-ID: 9ECF9E0004C, Message-ID: <AJYA0S.BEBA7LXKKXBE2@domain.com>, mail_id: 7EdJJKSLNSte, Hits: -, size: 1102, queued_as: 4956DE0004E, 348 ms
Sep  1 12:30:53 mail2 postfix/smtp-amavis/smtp[979]: 9ECF9E0004C: to=<recipient@hotmail.com>, relay=127.0.0.1[127.0.0.1]:10024, delay=0.98, delays=0.62/0.01/0.01/0.35, dsn=2.0.0, status=sent (250 2.0.0 from MTA(smtp:[127.0.0.1]:10025): 250 2.0.0 Ok: queued as 4956DE0004E)
Sep  1 12:30:53 mail2 postfix/qmgr[944]: 9ECF9E0004C: removed
Sep  1 12:30:54 mail2 postfix/smtp[983]: Trusted TLS connection established to hotmail-com.olc.protection.outlook.com[104.47.51.33]:25: TLSv1.2 with cipher ECDHE-RSA-AES256-GCM-SHA384 (256/256 bits)
Sep  1 12:30:55 mail2 postfix/smtp[983]: 4956DE0004E: to=<recipient@hotmail.com>, relay=hotmail-com.olc.protection.outlook.com[104.47.51.33]:25, delay=2.5, delays=0.23/0.05/0.64/1.6, dsn=2.6.0, status=sent (250 2.6.0 <AJYA0S.BEBA7LXKKXBE2@domain.com> [InternalId=854698510561, Hostname=AS8PR07MB9378.eurprd07.prod.outlook.com] 10353 bytes in 0.243, 41.484 KB/sec Queued mail for delivery -> 250 2.1.5)
Sep  1 12:30:55 mail2 postfix/qmgr[944]: 4956DE0004E: removed
```

**Comment by Encephala** [comment]
Awesome, good luck with the rest of the maintenance!
